### PR TITLE
GG-29644: DR: Incremental state transfer index.

### PR DESCRIPTION
**Pull request by AMashenkov**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/tree/AbstractUpdateLogInnerIO.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        int dstIdx,
+        BPlusIO<UpdateLogRow> srcIo,
+        long srcPageAddr,
+        int srcIdx) throws IgniteCheckedException {
+        int dstOff = offset(dstIdx);
+
+        long link = ((PendingRowIO)srcIo).getLink(srcPageAddr, srcIdx);
+        long expireTime = ((PendingRowIO)srcIo).getExpireTime(srcPageAddr, srcIdx);
+
+        PageUtils.putLong(dstPageAddr, dstOff, expireTime);
```
[the remark this answers is not in the document]
Comment (AMashenkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/tree/AbstractUpdateLogInnerIO.java:66]: I think "inlining" in this particular case doesn't mentioned in JMM spec.

Who will "inline" this in runtime? C1 or C2. Any proof?
What if user disable C2  due to any reasons?

**Review by korlov42 — APPROVED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/version/GridCacheVersionEx.java [hunk 2]
```diff
@@ -105,7 +107,7 @@ public GridCacheVersionEx(int topVer, int nodeOrderDrId, long order, GridCacheVe
         }
 
         switch (writer.state()) {
-            case 3:
+            case 4:
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/version/GridCacheVersionEx.java:110]: Hate to say this but I believe that new version of object won't be serialized properly on old nodes when RU enabled.
It'll try to read field "drVer" but "updateCounter" is found. "u" goes after "d" in alphabet, so reader will assume that "drVer" field is removed and will nullify it. Have you tested it? Am I wrong?
We need to make sure that this code is 100% safe.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/io/PagePartitionMetaIO.java [hunk 3]
```diff
@@ -46,7 +46,8 @@
     /** */
     public static final IOVersions<PagePartitionMetaIO> VERSIONS = new IOVersions<>(
         new PagePartitionMetaIO(1),
-        new PagePartitionMetaIOV2(2)
+        new PagePartitionMetaIOV2(2),
+        new PagePartitionMetaIOV2(Short.toUnsignedInt((short)-1)) // Prevent partition usage on old versions after upgrade.
```
[the remark this answers is not in the document]
Comment (AMashenkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/io/PagePartitionMetaIO.java:50]: Sure, we discussed this with Alex Goncharuk and concluded 
as we can't reserve version number in Ignite, then we will use negative versions in GG CE.
This looks a bit tricky due to lack of PageIO's flexibility.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java [hunk 4]
```diff
@@ -2031,53 +2077,55 @@ private Metas getOrAllocatePartitionMetas() throws IgniteCheckedException {
...
+                                        ", grpId=" + grpId + ", oldVer=" + pageVer +
+                                        ", newVer=" + io.getVersion()
                                 );
 
                             io = PagePartitionMetaIO.VERSIONS.latest();
 
-                            ((PagePartitionMetaIOV2)io).upgradePage(pageAddr);
+                            assert io.getVersion() == Short.toUnsignedInt((short)-1);
+
+                            ((PagePartitionMetaIOV2) io).upgradePage(pageAddr, pageVer);
```
[the remark this answers is not in the document]
Comment (AMashenkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheOffheapManager.java:2091]: Got it.
I'll create a separate class for new version.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/version/GridCacheVersionEx.java (hunk 2 above)
Comment (AMashenkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/version/GridCacheVersionEx.java:110]: Sounds sad. I believe it is safe to add new fields.
I'll check this.

Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/version/GridCacheVersionEx.java]: Cool, now it works for sure!

**Pull request merged**